vim-patch:8.0.1455,8.1.{2115,2361}

janlazo opened the pull request:
I don't understand why this patch necessary. The test doesn't check for regressions, especially for Windows shells.

janlazo committed:
vim-patch:8.1.2115: MS-Windows: shell commands fail if &shell contains a space

Problem:    MS-Windows: shell commands fail if &shell contains a space.
Solution:   Use quotes instead of escaping. (closes vim/vim#4920)
https://github.com/vim/vim/commit/2efc44b3f0b6bd8307cb281af095e08e15ab1c24

Always double-quote &shell if it contains a space.
Neovim does not support escaping space with backslash, unlike Vim.

N/A patches for version.c:

vim-patch:8.0.1455: if $SHELL contains a space then 'shell' is incorrect

Problem:    If $SHELL contains a space then the default value of 'shell' is
            incorrect. (Matthew Horan)
Solution:   Escape spaces in $SHELL. (Christian Brabandt, closes vim/vim#459)
https://github.com/vim/vim/commit/4bfa8af14142e54f509048239f4e8596911f56aa

vim-patch:8.2.1194: test failure because shell prompt differs

Problem:    Test failure because shell prompt differs.
Solution:   Set the shell prompt.
https://github.com/vim/vim/commit/a4dc6f92bb29b00783f8945bbe1101e837b6ad3c

blueyed commented on src/nvim/testdir/test_startup.vim:
```diff
@@ -499,6 +499,20 @@ func Test_read_stdin()
+  if RunVimPiped([], after, '', '')
+    let lines = readfile('Xtestout')
+    " MS-Windows adds a space after the word
+    call assert_equal('/bin/with\ space/sh', lines[0])
```
Doesn't this fail with the rest of the patch?

janlazo replied:
Try running the tests with `SHELL` having a space. Then, try it on Windows. The patch should have used double quotes instead in line with the suggestion in `:h shell`.

janlazo commented:
I prefer not merging this until it's fixed upstream. I tried setting `SHELL` to sh with a space in the filepath. `:!` bang commands don't work. Tested on `vim -u NONE`.

I leave it you guys to decide what to do with this patch. I always set `shell` in my vimrc to avoid these convenient hacks and I added a check today for `$SHELL` on `+unix` today.

blueyed commented:
/cc @chrisbra
Is the patch how you've meant it to be initially after all?

janlazo commented:
[vim-patch:8.1.2115](https://github.com/vim/vim/commit/2efc44b3f0b6bd8307cb281af095e08e15ab1c24) helps here on Windows. What's the problem with using double quotes for all OS in general?

janlazo commented:
Are there (neo)vim plugins that detect escaped `&shell` values? I don't know how to unescape `&shell` if this is merged.

justinmk commented:
> What's the problem with using double quotes for all OS in general?

No problem unless the path has a `\` in it, but that seems unlikely.

> I don't know how to unescape `&shell` if this is merged. Of course, it has to work on Windows too.

If double quotes solves that problem, let's just do double quotes.

janlazo commented:
Is wrapping the shell value in double quotes enough? That's what I do on my vimrc so that unescaping is easy: remove the double quotes.

justinmk commented:
> Is wrapping the shell value in double quotes enough? That's what I do on my vimrc so that unescaping is easy: remove the double quotes.

It can't be worse than the pile of hacks being augmented by this vim-patch. Let's just try it.

janlazo commented:
Ok. I'll get back to this after resolving `check.vim` for the upstream tests. I'll update the docs and tests to use double quotes for escaped `$SHELL`. The notable source change for 8.0.1455, `set_string_default_esc`, will be reverted in favor of the double-quote approach in 8.1.2115 and 8.1.2116.

That said, I don't know which parts of source code require `&shell` to be escaped. `:terminal` doesn't work on Windows if `&shell` is an absolute filepath. Tested with `$COMSPEC`.

chrisbra commented:
> Is the patch how you've meant it to be initially after all?

Not sure I understand the question. Initially I just escaped any spaces in the shell option, but that might break when giving arguments to the shell option (or using `job_start()`). So not sure what it should look like completely. The mentioned vim patch was done by Bram, not by me, because my patch had some problems.

Note, there is also https://github.com/vim/vim/pull/4920 to fix a similar problem.

justinmk commented:
How does this (or upstream Vim) deal with cases like this: https://github.com/tpope/vim-sensible/blob/4a7159a30061b26aec9ac367823094e7fd0f6a5b/plugin/sensible.vim#L72

    set shell=/usr/bin/env\ bash

There 'shell' has a space but it should not be quoted.

janlazo committed:
vim-patch:8.1.2361: MS-Windows: test failures related to VIMDLL

Problem:    MS-Windows: test failures related to VIMDLL.
Solution:   Adjust code and tests. (Ken Takata, closes vim/vim#5283)
https://github.com/vim/vim/commit/310c32e8920140f0db747c6c6eb06b1ee53cdb5a

janlazo commented:
> How does this (or upstream Vim) deal with cases like this: https://github.com/tpope/vim-sensible/blob/4a7159a30061b26aec9ac367823094e7fd0f6a5b/plugin/sensible.vim#L72
> 
> ```
> set shell=/usr/bin/env\ bash
> ```
> 
> There 'shell' has a space but it should not be quoted.

It doesn't. Tested on vim 8.2.1139 (appimage). I didn't add support for it here. Need a separate patch for that.
